### PR TITLE
HW35 Viktar

### PR DESCRIPTION
**Pull request by Amicronpersej8**

### CODE REVIEW

File: homework_35.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        name: "Henry Ford",
+        phone: "(09) 999 93 23",
+        email: "ford0@gmail.com",
+        cars: ["bmw", "audi"],
+        hasChildren: true,
+        hasEducation: false
+    }
+];
+// 1. Создать строку из имен пользователей через запятую
+var getUsernameString = function (array) {
```
Comment (jathpr) [inline homework_35.js:28]: prefer explicit names as users against common array

File: Homework_35.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    // 2. Посчитать общее количнство машин у пользователей
+    const getAmountCars = (array: Users[]) => {
+        const arrCars = []
+        array.forEach(element => {
+            if (element.cars) {
+                arrCars.push(element.cars)
+            }
+        })
+        const resultArr = []
+        return resultArr.concat(...arrCars).length
```
Comment (jathpr) [inline Homework_35.ts:57]: good place for reduce and [...acc, ...cars]

File: Homework_35.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    }
+        console.log(getAmountCars(users), 'cars')
+        
+        // 3. Создать функцию, которая бы принимала массив пользователей и отфильтровывала пользователей на наличие образования
+    const getUsersWithEducation = (array: Users[]) => {
+        const nextUsers = []
+        array.filter(item => {if (item.hasEducation === true){
```
Comment (jathpr) [inline Homework_35.ts:64]: item.hasEducation && nextUsers.push(item.name)

File: Homework_35.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        nextUsers.push(item.name)}
+        })
+        return nextUsers
+    }    
+    console.log(getUsersWithEducation(users), 'with education')
+
+        // 4. Создать функцию, которая бы принимала массив пользователей и отфильтровывала пользователей на наличие животных
+        const getUsersWithAnimals = (array: Users[]) => {
+            const usersWithAnimals = []
+            array.filter(item => {if (item.animals){
```
Comment (jathpr) [inline Homework_35.ts:74]: misuse of filter
`usersWithAnimals = users.filter(user => user.animals)`

File: Homework_35.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                usersWithAnimals.push(item.name)}
+            })
+            return usersWithAnimals
+        }    
+        console.log(getUsersWithAnimals(users), 'with animals')
+
+        // 5. Создать функцию, которая бы принимала массив пользователей и отдавала бы  строку с названиями марок автомобилей через запятую
+        const getCars = (array: Users[]) => {
+            const arrCars = []
+            array.map(element => {
```
Comment (jathpr) [inline Homework_35.ts:84]: misuse of map, you need foreach

File: Homework_35.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                if (element.cars) {
+                    arrCars.push(element.cars)
+                }
+            })
+            return arrCars.toString()
```
Comment (jathpr) [inline Homework_35.ts:89]: arrCars is array of arrays, you need array of strings

**Review by jathpr — CHANGES_REQUESTED** [review]
fine, but need fixes

File: Homework_35.ts (hunk 6 above)
Comment (Amicronpersej8) [reply Homework_35.ts:89]: arrCars.toString() makes string. Is this wrong? The task says: "принимала массив пользователей и отдавала бы  строку с названиями марок автомобилей"

File: Homework_35.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  return usernameStr;
+};
+console.log(getUsernameString(users));
+
+// 2. Посчитать общее количнство машин у пользователей
+const getAmountCars = (users: Users[]) => {
+  const arrCars = [];
+  users.forEach((user) => {
+    if (user.cars) {
+      arrCars.push(user.cars);
```
Comment (jathpr) [inline Homework_35.ts:51]: you can try `arrCars.push(...user.cars);`
or smthng similar if it doesn't work

File: Homework_35.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+// 5. Создать функцию, которая бы принимала массив пользователей и отдавала бы  строку с названиями марок автомобилей через запятую
+const getCars = (array: Users[]) => {
+  const arrCars = [];
+  array.forEach((user) => {
+    if (user.cars) {
+      arrCars.push(user.cars);
+    }
+  });
+  return arrCars.toString();
```
Comment (jathpr) [inline Homework_35.ts:76]: arrCars - is array of arrays, not a array of string
it returns string with `[]` in it, this is a mistake

**Comment by jathpr** [comment]
fix the last task and do not forget to close resolved conversations

File: Homework_35.ts (hunk 8 above)
Comment (Amicronpersej8) [reply Homework_35.ts:74]: By the way,  method array.toString() array of arrays convert to string

Comment (jathpr) [reply Homework_35.ts:74]: thanks, I will check

Comment (jathpr) [reply Homework_35.ts:74]: You are right! So it works

But still - `bmw,bmw,audi` - looks not perfect, and it's hard to change it
On prod - I wouldn't get it, but here it's ok.

**Pull request merged**